### ISSUE TITLE
Return value for `atoms=true` should be whole numbers

### ISSUE THREAD

**Issue by fleupold** [gnosis/dex-services#1111]
> https://price-estimate-mainnet.dev.gnosisdev.com/api/v1/markets/7-4/estimated-amounts-at-price/0.000000000001019854?atoms=true
{"baseTokenId":"7","quoteTokenId":"4","buyAmountInBase":"25323216136017510000000","sellAmountInQuote":"25825983269.182007"}

Note, that sellAmountInQuote is returned as a fraction of atoms. Semantically, I don't think it makes sense to receive a fraction of atoms (which by definition should be the smallest indivisible unit). I'd therefore suggest to floor this number and make sure we are returning "whole" numbers.

**Comment by fleupold** [comment]
@e00E mentioned on slack:

> Since we are returning json I prefer this being a json number which can be fractional. Even with atoms=true with things like the rounding buffer a fractional value is sensible.

**Comment by fleupold** [comment]
I don't see how the rounding buffer makes fractional atoms more sensible. As a consumer of this API I want to know how many atoms I can sell or buy (in most cases to interact with the Smart Contract) and I shouldn't have to know anything about what or how rounding is applied (all I know is I can somewhat "trust" this number to give me a trade).
If I receive a fractional result I don't see any way other than rounding down on the client on how to convert this information into something I can actually make use of.

**Comment by e00E** [comment]
In euros the atom is a cent. If I'm selling something at a price of 0.5 cents and you ask me how much you need to pay to receive one item then answer is 1 cent. Yet if I answer 0.5 this helps you realize that you could also buy 2 for the same price.

Regardless of the answer to whether we should round when we're parsing the number in another application we still won't be able to do in rust `u128::from_str` because we encode the number according to json which means it could look like `10.0` or `1e1`.

**Comment by fleupold** [comment]
I agree that the price can remain in floats, however the amounts you receive/sell in atoms shouldn't be fractional. Note, that due to slippage the doubling of sell amounts doesn't mean a doubling of buy amounts per se.

As for the implementation, I think the string should never have float notation (scientific notation or trailing .0). Maybe we cannot use `display_fromstr` in that case although it seems to work for now. To be more robust we could have an opaque field atoms on `EstimatedOrderResult` struct that doesn't get serialized but can be used in a custom serialization to convert to uint before converting to string.

**Comment by e00E** [comment]
> As for the implementation, I think the string should never have float notation

I don't have a strong opinion on whether we floor but this I strongly disagree with. We are returning json numbers (well, we would be if we changed the numbers from being a string to actual numbers which is a change we want to do) and in json the numbers `1` `1.0` and `1e0` all refer to the same number "one". It don't see why we would arbitrarily decide to only allow one of those versions.

**Comment by fleupold** [comment]
`/markets` return prices (which make sense to be float) and `/estimated-buy-amount` return amounts. If `atoms=false` this amount also makes sense to be a fraction. 
If `atoms=true` however we deal with on average 18 orders of magnitude larger number in which case precision becomes an issue.

Maybe the difference between `atoms=true|false` is indeed more substantial than we originally thought and should be represented by having its own path (which would then also warrant a different return type structure). This way we could change estimated-buy-amount to return float but `estimated-buy-amount-atoms` would keep returning a whole number string.

**Comment by nlordell** [comment]
> As for the implementation, I think the string should never have float notation (scientific notation or trailing .0).

I also agree with @e00E, since `1.0`, `1e1` and `10` are all valid representations for the exact same JSON object as far as EcmaScript/JSON spec is concerned. That would mean we would rely on specific implementation details of the JSON serialization library.

One proposal would be to have an intermediate struct:
```
enum Amount {
    Atom(u128),
    Base(f64),
}
```

to represent amounts in the API results, such that serialization looks like: `Amount::Atom(100) => "100"` and `Amount::Base(100) => 100`.

> In euros the atom is a cent. If I'm selling something at a price of 0.5 cents and you ask me how much you need to pay to receive one item then answer is 1 cent. Yet if I answer 0.5 this helps you realize that you could also buy 2 for the same price.

I don't think this is true. While I understand the argument that we are losing information, I think that the information being lost has no value. Atoms can only be whole numbers, so it is super unclear what a "half-atom" actually means. Additionally, prices don't scale linearly (because of slippage) so just because you can sell 1 DAI atom for 0.99 PAX atoms, does not mean you can sell 1e18 DAI atoms for 0.99 * 1e18 PAX atoms.

I think if we really wanted to express prices, then we should have an API that returns the ratio instead of one that computes buy and sell amounts in atoms.

Furthermore, I think the rounding imprecision is a product of internal implementation details (for example, if we implemented the `pricegraph` with real numbers instead of floats, then it would be able to return integer values instead). Ultimately, I also believe that `Pricegraph::order_for_sell_amount` should return a `TransitiveOrder` with rounding applied, so I think this shouldn't be fixed in the service either.

**Comment by e00E** [comment]
In practice, because pricegraph uses f64 internally we are not going actually use higher precision even if we serialize atomic amounts as u128 strings. For this reason it makes sense to have all numbers be real json numbers instead of strings. This also makes them the most easy to work with in client code. We can still document in the api that atoms are going to be natural numbers.

In a way, this even forward compatible with a pricegraph change that would allow for higher precision by relying on json parsers that are able to handle numbers outside the f64 range. For example I believe serde_json can interpret numbers as strings so that you can deserialize them into some BigReal or BigInteger type.

**Comment by nlordell** [comment]
> In a way, this even forward compatible with a pricegraph change that would allow for higher precision by relying on json parsers that are able to handle numbers outside the f64 range.

In practice this is not true. Especially since one of the parsers we will use with the service is the ECMAScript `JSON.parse` which explicitly uses the the ECMA `Number` type which is a `f64/u32`.

>  For example I believe serde_json can interpret numbers as strings so that you can deserialize them into some BigReal or BigInteger type.

This requires the `arbitrary-precision` feature which breaks all sorts of things (and we have had issues in the past because of untagged enums): https://github.com/serde-rs/json/issues?q=is%3Aissue+is%3Aopen+arbitrary

If we want to represent an `f64` or `u128` (or BigInt), then they require two different representations and we cannot rely on JSON numbers.

**Comment by e00E** [comment]
I quickly found https://github.com/josdejong/lossless-json for javascript which parses json numbers as strings so that they can be turned into for example big integers.

For serde_json you probably know the situation better than I do but there is a workaround listed in the issue or maybe we can fix it in serde upstream or it won't be a problem with the simple return types in the price estimator or we can use another json library.

While I would like to use json numbers for niceness I do see the practical issues you raise. If we do want to encode amounts as strings then we should use this encoding for amounts everywhere, even in `/markets` so that we are consistent and forward compatible with a pricegraph implementation with higher precision. In this case I would describe the format in our readme as:

> A json string whose contents are a json number. This number is not limited to the range of double precision floats as is the case for json numbers in many json implementations.

**Comment by nlordell** [comment]
We could use an untagged enum:
```rust
enum Amount {
    Atom(u128),
    Base(f64),
}
```

**Comment by e00E** [comment]
If that leads to following the format I describe at the end of my previous post (I think it does) then sure. I just don't want to mix formatting amounts in some places as json numbers and in other places as json strings that represent numbers in our output.

**Comment by nlordell** [comment]
> there is a workaround listed in the issue

The issue is that Cargo uses one feature set per crate. This means that it affects serialization of other types outside of the `dex-services` repo, specifically this one we have had issues with in the past: https://docs.rs/jsonrpc-core/14.2.0/jsonrpc_core/types/id/enum.Id.html

> or maybe we can fix it in serde upstream

I briefly tried to do this once, but its a pretty complicated issue since `Serde` serialization types themselves don't support arbitrary precision integers so the fix gets kind of messy.

> or it won't be a problem with the simple return types in the price estimator or we can use another json library.

That is an option.

**Comment by nlordell** [comment]
> If that leads to following the format I describe at the end of my previous post 

I meant it more like:
```
// atoms=false
{ sellAmoutInBase: "1000000000000000000", ...}
// atoms=true
{ sellAmoutInBase: 1.0, ...}
```

**Comment by e00E** [comment]
If the client has write code to extract the number anyway then it seems easier to me if it's always a string. Especially because there can still be precision issues with atoms=true like a token having a decimals==1 or really large buy/sell amounts for whatever reason. And if the client doesn't want to use some BigInteger they can still parse the string as a float which will presumably go to the closest floating point number.

**Comment by nlordell** [comment]
> If the client has write code to extract the number anyway then it seems easier to me if it's always a string.

I think having to handle the data different for these two cases is fine, especially since the data represents two fundamentally different types. This way clients are forced to handle `atoms=true` and `atoms=false` differently, which, while it may be annoying, requires explicit conversion between the units.

**Comment by nlordell** [comment]
If we want to be more explicit, then I think we should use a tagged unit so:
```
// atoms=false
{ sellAmoutInBase: { unit: "atoms", value: "1000000000000000000" }, ...}
// atoms=true
{ sellAmoutInBase: { unit: "base_unit", value: 1.0 }, ...}
```

**Comment by e00E** [comment]
What I'm trying to say is that we want to have the potential for higher precisions for all routes amounts anyway. There is no drawback in making this a string too. The only change for clients is that they'll have to do the same string to number conversion for /markets too that they already do for the other routes and then we are forward compatible with any pricegraph change that would give us more precision.
The enum approach isn't bad, it's just using the string everywhere is even better.

**Comment by nlordell** [comment]
> There is no drawback in making this a string too.

I think its counter-intuitive. My main concern, however, is that it doesn't fully capture the fact that the data encoded in the string may have different types. If its a string, should you decode it as a big integer or a 64-bit float? Is the value represented in atoms or in base units? This makes it harder to consume the data in a vacuum (without knowledge of the `atoms` flag, if it was saved to file by some service for example.

Also serializing it with a `string | number` instead of `string` lets us stay backwards compatible, which is also a :+1: .

**Comment by e00E** [comment]
There a couple of related things here:

Do we want to be able to support numbers with precision higher than f64?
The answer is definitely yes for atoms=false and as I have argued alsob yes for atoms=true.

Should the higher precision number type be a string or a json number?
Initially I argued yes but based on your arguments it should probably be a string.

Should `/markets` be high precision or f64?
This route should work the same way as the other routes. If that means not being backwards compatible then so be it. Better to fix the api than leave it wrong. This means that the "volume" and "price" here should be able to be the high precision type too despite not being "atoms" or "baseunit".

At this point according to my answers all the json numbers (except token ids) should be strings because they all benefit from potentially higher precision.

Now we can introduce more type safety:

Should we have an extra type for amounts that cannot be fractional?
Based on every number benefitting from high precision I find this doesn't add value over just writing in the api docs that atoms will always be non fractional numbers.

Should we have an extra type to distinguish atoms and baseunit?
This would automatically also solve the answer the previous question because the atoms type is non fractional. I don't find this useful because a user is supposed to read the docs and decide whether they have set atoms to true or false. And if they ignore that then they can also ignore the extra type safety here. I don't like parroting back what was already specified in the request as I feel it just makes the type more cumbersome to deal with.

**Comment by nlordell** [comment]
> Better to fix the api than leave it wrong.

I also agree with this, but not having to is a nice bonus :smile: .

> I have argued alsob yes for atoms=true.

Sorry, I had missed this point. Yes in that case we would not be able to use JSON numbers to represent these values.

> Should /markets be high precision or f64? [...] Should we have an extra type for amounts that cannot be fractional?

The part I'm struggling with is that for me we have really 3 possible types now:
- `f64` - JSON Numbers
- `u128` - JSON Strings (for `volume`s in atoms and but **not** `volume`s in base units and `price`s).
- `Real` - JSON Strings (for `volume`s in base units and `price`s).

Currently, we arguable only use two of them: `f64` for `volume`s in base units and `price`s and `u128` for `volume`s in atoms. I think we should design the API given what we currently use. If we increase the precision to use `Real`s we can re-evaluate.

Regarding the statement that we use `u128` for `volume`s in atoms, this is a more of a conceptual thing in that these amounts represent atoms that cannot be split into smaller quantities, and as such need to be whole numbers.

**Comment by nlordell** [comment]
In my mind, we currently have:
- `price`s and `amount`s in base units are `f64`, as this is the natural way to represent these values, and how they are currently represented internally
- `amount`s in atoms are `u128`, albeit being *internally* represented as `f64`s as an approximation.

This makes it, for me, the most logical to use `string | number` since:
- It accurately represents both of these types
- It keeps us backwards compatible (as a bonus).
- Doesn't represent `f64`s as strings which is very counter-intuitive and not really standard practice for JSON (I think it will be more confusing than anything).

If we decide to implement arbitrary precision for `price`s, then we can re-evaluate and introduce a breaking API change, but I really don't think it is necessary at the moment.

**Comment by nlordell** [comment]
Additionally, note that API breaking changes are harder to deal with in our case, because of how Mesa is released and managed by the dxDAO, so while its not ideal to give so much weight to maintaining backwards compatibility at a philosophical level, I think for our specific case it does have some value.

**Comment by fleupold** [comment]
> If we want to be more explicit, then I think we should use a tagged union (also a fairly common pattern in JS/JSON) so:

I actually do like that approach (it would also alleviate the concerns of https://github.com/gnosis/dex-services/issues/1106). I think just having a `string | number` encoding is a bit too subtle (it's easy to miss the difference).

> In my mind, we currently have: [...] amounts in atoms are u128

I thought so too but unfortunately the volume in the [markets endpoint](https://dex-price-estimator.gnosis.io//api/v1/markets/1-0?atoms=true) is encoded as a float as well even if `atoms=true`

**Comment by nlordell** [comment]
> I think just having a string | number encoding is a bit too subtle (it's easy to miss the difference).

I think that is true for TS/JS but for strongly typed languages I think it is less subtle. For example in Go, C#, or other languages that use reflection for JSON parsing, this would be more immediately apparent.

> I thought so too but unfortunately the volume in the markets endpoint is encoded as a float 

Now that we have discussed it a bit more, I think the more correct approach would be to use a string representation of `u128`s.

**Comment by nlordell** [comment]
Closed by #1202, Created #1233 for outstanding discussion.
